rdt/hyper_transformer.py

```diff
@@ -80,73 +78,62 @@ class HyperTransformer:
-                Data used to analyze the ``pandas.DataFrame`` dtypes.
-
-        Returns:
-            dict:
-                Mapping of column names and transformer instances.
+        self.field_types = field_types or {}
+        self._default_transformers = get_default_transformers()
+        self.data_type_transformers = data_type_transformers or {}
+        self.field_transformers = field_transformers or {}
+        self.transform_output_types = transform_output_types or self._DEFAULT_OUTPUT_TYPES
 
-        Raises:
-            ValueError:
-                if a ``dtype`` is not supported by the `HyperTransformer``.
-        """
-        transformers = {}
-        if self.dtypes:
-            dtypes = self.dtypes
-        else:
-            dtypes = [
-                data[column].dropna().infer_objects().dtype
-                for column in data.columns
-            ]
-
-        for name, dtype in zip(data.columns, dtypes):
-            try:
-                kind = np.dtype(dtype).kind
-            except TypeError:
-                # probably category
-                kind = 'O'
-
-            transformer_template = self.dtype_transformers.get(kind)
-            if not transformer_template:
-                raise ValueError('Unsupported dtype: {}'.format(dtype))
-
-            if isinstance(transformer_template, str):
-                transformer_template = self._TRANSFORMER_TEMPLATES[transformer_template]
-
-            if not isinstance(transformer_template, type):
-                transformer = deepcopy(transformer_template)
+    @staticmethod
+    def _field_in_data(field, data):
+        all_columns_in_data = isinstance(field, tuple) and all(col in data for col in field)
+        return field in data or all_columns_in_data
+
+    def _update_field_types(self, data):
+        # get set of provided fields including multi-column fields
+        provided_fields = set()
+        for field in self.field_types.keys():
+            if isinstance(field, tuple):
+                provided_fields.update(field)
             else:
-                transformer = transformer_template()
-
-            transformers[name] = transformer
-
-        return transformers
+                provided_fields.add(field)
+
+        for field in data:
+            if field not in provided_fields:
+                self.field_types[field] = self._DTYPES_TO_DATA_TYPES[data[field].dtype.kind]
+
+    def _fit_field_transformer(self, data, field, transformer):
+        transformer = load_transformer(transformer)
+        transformer.fit(data, field)
+        self._transformers_sequence.append(transformer)
+        output_types = transformer.get_output_types()
+        next_transformers = transformer.get_next_transformers()
+        for (output, output_type) in output_types.items():
+            if output_type in self.transform_output_types:
+                self._output_columns.append(output)
+            else:
+                self._generated_columns.append(output)
```

Comment:
Do we need this? I thought we only needed `_input_columns` and `_output_columns`

Reply:
My understanding is that if they transform a table with new columns, we don't want to do anything to those. However, how do we distinguish new columns from generated columns? Generated columns aren't in the input or output so I couldn't think of another way to filter them out

Reply:
Oh! Of course! You mean columns that were generated in an intermediate step, but which were further transformed by another transformer before finishing, right?